src/common/src/session_config/mod.rs

```diff
@@ -12,1082 +12,284 @@
-type RowSecurity = ConfigBool<ROW_SECURITY, true>;
-type StandardConformingStrings = ConfigString<STANDARD_CONFORMING_STRINGS>;
-type StreamingRateLimit = ConfigU64<STREAMING_RATE_LIMIT, 0>;
-type CdcBackfill = ConfigBool<CDC_BACKFILL, false>;
-type BackgroundDdl = ConfigBool<BACKGROUND_DDL, false>;
-type ServerEncoding = ConfigString<SERVER_ENCODING>;
-type StreamingEnableArrangementBackfill = ConfigBool<STREAMING_ENABLE_ARRANGEMENT_BACKFILL, false>;
-
-/// Report status or notice to caller.
-pub trait ConfigReporter {
-    fn report_status(&mut self, key: &str, new_val: String);
-}
-
-// Report nothing.
-impl ConfigReporter for () {
-    fn report_status(&mut self, _key: &str, _new_val: String) {}
-}
-
-#[derive(Educe)]
-#[educe(Default)]
+/// This is the Session Config of RisingWave.
+#[derive(SessionConfig)]
 pub struct ConfigMap {
     /// If `RW_IMPLICIT_FLUSH` is on, then every INSERT/UPDATE/DELETE statement will block
     /// until the entire dataflow is refreshed. In other words, every related table & MV will
     /// be able to see the write.
-    implicit_flush: ImplicitFlush,
+    #[parameter(default = false, rename = "RW_IMPLICIT_FLUSH")]
+    implicit_flush: bool,
 
     /// If `CREATE_COMPACTION_GROUP_FOR_MV` is on, dedicated compaction groups will be created in
     /// MV creation.
-    create_compaction_group_for_mv: CreateCompactionGroupForMv,
+    #[parameter(default = false)]
+    create_compaction_group_for_mv: bool,
 
     /// A temporary config variable to force query running in either local or distributed mode.
     /// The default value is auto which means let the system decide to run batch queries in local
     /// or distributed mode automatically.
+    #[parameter(default = QueryMode::default())]
     query_mode: QueryMode,
 
-    /// see <https://www.postgresql.org/docs/current/runtime-config-client.html#:~:text=for%20more%20information.-,extra_float_digits,-(integer)>
-    extra_float_digit: ExtraFloatDigit,
+    /// Sets the number of digits displayed for floating-point values.
+    /// See <https://www.postgresql.org/docs/current/runtime-config-client.html#:~:text=for%20more%20information.-,extra_float_digits,-(integer)>
+    #[parameter(default = 1)]
+    extra_float_digits: i32,
 
-    /// see <https://www.postgresql.org/docs/14/runtime-config-logging.html#:~:text=What%20to%20Log-,application_name,-(string)>
-    application_name: ApplicationName,
+    /// Sets the application name to be reported in statistics and logs.
+    /// See <https://www.postgresql.org/docs/14/runtime-config-logging.html#:~:text=What%20to%20Log-,application_name,-(string)>
+    #[parameter(default = "", flags = "REPORT")]
+    application_name: String,
 
+    /// It is typically set by an application upon connection to the server.
     /// see <https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-DATESTYLE>
-    date_style: DateStyle,
+    #[parameter(default = "", rename = "DATESTYLE")]
+    date_style: String,
```

Comment:
There are some variables that:
* Defaults to a hardcoded default value.
* When explicitly set to this value, do nothing.
* When explicitly set to other values, report unsupported error.

For example
* `DateStyle` = `ISO, MDY`
* `IntervalStyle` = `postgres`
* `bytea_output` = `hex`
* `client_encoding` = `UTF8`

It may be more convenient to have `check_hook = check_allowed(&["UTF8"])`

Reply:
will add in next PR

Reply:
check_hook covered this case already, this check_allowed is for easy use. 
```

fn check_timezone(val: &str) -> Result<(), String> {
    // Check if the provided string is a valid timezone.
    Tz::from_str_insensitive(val).map_err(|_e| "Not a valid timezone")?;
    Ok(())
}

fn check_client_encoding(val: &str) -> Result<(), String> {
    // https://github.com/postgres/postgres/blob/REL_15_3/src/common/encnames.c#L525
    let clean = val.replace(|c: char| !c.is_ascii_alphanumeric(), "");
    if !clean.eq_ignore_ascii_case("UTF8") {
        Err("Only support 'UTF8' for CLIENT_ENCODING".to_string())
    } else {
        Ok(())
    }
}

fn check_bytea_output(val: &str) -> Result<(), String> {
    if val == "hex" {
        Ok(())
    } else {
        Err("Only support 'hex' for BYTEA_OUTPUT".to_string())
    }
}
```